src/v/cluster/producer_state.h

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/types.h"
+#include "model/record.h"
+#include "utils/intrusive_list_helpers.h"
+#include "utils/mutex.h"
+#include "utils/rwlock.h"
+
+#include <seastar/core/shared_future.hh>
+#include <seastar/util/defer.hh>
+
+// Befriended to expose internal state in tests.
+struct test_fixture;
+
+namespace cluster {
+
+template<class Func>
+concept AcceptsUnits = requires(Func f, ssx::semaphore_units units) {
+    f(std::move(units));
+};
+
+class producer_state_manager;
+class producer_state;
+struct producer_state_snapshot;
+class request;
+
+using producer_ptr = ss::lw_shared_ptr<producer_state>;
+using result_promise_t = ss::shared_promise<result<kafka_result>>;
```

Comment:
whenever i see a `shared_promise` i am wondering if we can replace it with condition variable and a value holder. For me the shared promise semantics is unclear

Reply:
I like `shared_promise` because it already holds a value and we don't need extra code to support the holder. Which part of the `shared_promise` confuses you?

Reply:
Was chatting with Michal earlier today, he was referring to this behavior of shared_promise which is unintuitive that he ran into.

```
    ss::shared_promise<int32_t> promise;
    auto future = promise.get_shared_future();
    BOOST_REQUIRE_EQUAL(promise.available(), false);
    promise.set_value(42);
    BOOST_REQUIRE_EQUAL(promise.available(), false);
```

available() is false right after set_value(). I dug a little into the implementation, think thats because of the implementation of then_wrapped that moves the future state into the continuation. shared_future::get_future does this to chain the result of shared futures with the original future.

```
                if (_original_future._state.valid()) {
                    // _original_future's result is forwarded to each peer.
                    (void)_original_future.then_wrapped([s = this->shared_from_this()] (future_type&& f) mutable {
                    <---- available() unreliable here ---->
                        s->resolve(std::move(f));
                    });
                }


        explicit shared_state(future_type f) noexcept : _original_future(std::move(f)) { }
        void resolve(future_type&& f) noexcept {
            _original_future = std::move(f);  <---- set again
            auto& state = _original_future._state;
```

if we check available() before the continuation runs, it returns false because `f` is moved from _original_future. (resolve() sets it back, which is weird).

We can also reproduce this with a regular future.

```
SEASTAR_THREAD_TEST_CASE(future) {
    ss::promise<int> foo;
    auto f = foo.get_future().then_wrapped([] (auto&& result) {
        return result;
    });
    foo.set_value(1);
    BOOST_REQUIRE_EQUAL(foo.get_future().available(), false);
```

tldr is we cannot rely on shared_promise::available() being set right after set_value(), as long as we don't rely on that pattern, we are good here.

Reply:
You should add this excellent comment on the code.

Reply:
we have `available_promise` as a workaround; we may do the same for shared promise